mesonbuild/interpreterbase/interpreterbase.py

```diff
@@ -105,8 +123,13 @@ def load_root_meson_file(self) -> None:
         assert isinstance(code, str)
         try:
             self.ast = mparser.Parser(code, mesonfile).parse()
+            self.handle_meson_version_from_ast()
```

Comment:
The tests fail because this can only run during the main interpreter, not the AST one. My new plan is to have `handle_meson_version_from_ast()` in both the interpreterbase and the interpreter, the former will be a no-op return, the latter will parse the version and maybe error out.

Not sure what the best way of moving that into the parser is, but maybe it makes more sense to have the parser also return a list of FeatureCheck instances and invoke `.use()` from the interpreter.

Aside for that, does this look good?

Reply:
> Aside for that, does this look good?

I like the feature, but I'm not sure about implementation. It's maybe just nitpicking, but I'm not a big fan of parsing the file twice (even if it's only in error handling case) and that `strict` flag that does not really tell what it does. Feels better to handle the meson_version kwarg directly in mparser. In e7() just after `left = FunctionNode()` you can add `if left.func_name == 'project'` and do it there.

Are there any downside in that approach?

Reply:
Yes, we're pushing more interpretation into the AST builder. This PR already fails CI because `meson rewrite` cannot and should not do such interpretation, only parsing.

And then we also want to perform actions based on the self.subproject as well, so that's more information being passed from the interpreter into the AST builder.

It's a layering violation and it makes me uncomfortable.

Reply:
> It's a layering violation and it makes me uncomfortable.

Fair enough.

Alternatively, ParseException could contain the current AST. That way you don't have to re-parse and you don't need the `strict` flag.

Reply:
That's... actually a really good point, let me try to implement that.

Reply:
Played with the idea, what about https://github.com/mesonbuild/meson/compare/master...xclaesse:meson:mparse-ast-exception

Reply:
> Played with the idea, what about [master...xclaesse:meson:mparse-ast-exception](https://github.com/mesonbuild/meson/compare/master...xclaesse:meson:mparse-ast-exception)

Added your commit on top now, so that branch should be ready.